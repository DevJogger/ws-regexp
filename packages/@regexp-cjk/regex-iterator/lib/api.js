"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitLimit = exports.replaceEach = exports.reduce = exports.map = exports.each = void 0;
const each_1 = __importDefault(require("./api/each"));
exports.each = each_1.default;
const map_1 = __importDefault(require("./api/map"));
exports.map = map_1.default;
const reduce_1 = __importDefault(require("./api/reduce"));
exports.reduce = reduce_1.default;
const replaceEach_1 = __importDefault(require("./api/replaceEach"));
exports.replaceEach = replaceEach_1.default;
const splitLimit_1 = __importDefault(require("./api/splitLimit"));
exports.splitLimit = splitLimit_1.default;
//# sourceMappingURL=api.js.map