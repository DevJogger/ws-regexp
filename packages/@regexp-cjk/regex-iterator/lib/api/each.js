"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.each = void 0;
const _each_1 = require("./_each");
function each(input, re, cb, options = {}) {
    for (const m of _each_1._each(input, re, options)) {
        cb(m.match, m.index, m.re, m);
    }
}
exports.each = each;
exports.default = each;
//# sourceMappingURL=each.js.map