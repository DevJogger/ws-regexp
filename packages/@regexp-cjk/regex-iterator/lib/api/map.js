"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.map = map;
const _each_1 = require("./_each");
function _cb(match, index, re, _) {
    return match;
}
function map(input, re, cb, options = {}) {
    let ret = [];
    cb = cb !== null && cb !== void 0 ? cb : _cb;
    for (const m of (0, _each_1._each)(input, re, options)) {
        ret.push(cb(m.match, m.index, m.re, m));
    }
    return ret;
}
exports.default = map;
//# sourceMappingURL=map.js.map