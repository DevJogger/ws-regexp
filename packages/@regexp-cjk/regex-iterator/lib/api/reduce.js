"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reduce = exports._reduceCore = void 0;
const handleChainInput_1 = require("../util/handleChainInput");
const _each_1 = require("./_each");
const handleChainRowResult_1 = require("../util/handleChainRowResult");
function* _reduceCore(input, chain, options = {}) {
    let { allowFallbackToSource } = options;
    chain = chain.slice();
    const { regexp: re, backref } = handleChainInput_1.handleChainInput(chain.shift(), options);
    const options2 = {
        ...options,
        global: false,
    };
    for (const m of _each_1._each(input, re, options)) {
        let str = handleChainRowResult_1.handleChainRowResult({
            regexp: re,
            backref,
            match: m.match,
            allowFallbackToSource,
        }, m.match[0]);
        if (chain.length > 0) {
            for (const row of chain) {
                let { regexp, backref } = handleChainInput_1.handleChainInput(row, options2);
                const match = str.match(regexp);
                str = handleChainRowResult_1.handleChainRowResult({
                    regexp,
                    backref,
                    match,
                    allowFallbackToSource,
                }, str);
            }
        }
        yield str;
    }
}
exports._reduceCore = _reduceCore;
function reduce(input, chain, options = {}) {
    return [..._reduceCore(input, chain, options)];
}
exports.reduce = reduce;
exports.default = reduce;
//# sourceMappingURL=reduce.js.map