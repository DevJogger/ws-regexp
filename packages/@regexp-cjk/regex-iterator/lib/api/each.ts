import { IReturnTypeEachCore} from '../types';
import { _each } from './_each';
import { ICloneRegexpOptionsCustom } from '@regexp-cjk/clone-regexp/lib/types';

export function each<T extends RegExp = RegExp>(input: string,
	re: T,
	cb: (match: RegExpMatchArray, index: number, re: T, _: IReturnTypeEachCore<T>) => void,
	options: ICloneRegexpOptionsCustom<T> = {},
)
{
	for (const m of _each(input, re, options))
	{
		cb(m.match, m.index, m.re, m)
	}
}

export default each
