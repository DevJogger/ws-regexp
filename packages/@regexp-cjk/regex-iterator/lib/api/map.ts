import { IReturnTypeEachCore, ICallback } from '../types';
import { _each } from './_each';
import { ICloneRegexpOptionsCustom } from '@regexp-cjk/clone-regexp/lib/types';

function _cb<T extends RegExp = RegExp>(match: RegExpMatchArray, index: number, re: T, _: IReturnTypeEachCore<T>)
{
	return match
}

export function map<R = RegExpMatchArray, T extends RegExp = RegExp>(input: string,
	re: T,
	cb?: never,
	options?: ICloneRegexpOptionsCustom<T>,
): R[]
export function map<R, T extends RegExp = RegExp>(input: string,
	re: T,
	cb: ICallback<R, T>,
	options?: ICloneRegexpOptionsCustom<T>,
): R[]
export function map<R = RegExpMatchArray, T extends RegExp = RegExp>(input: string,
	re: T,
	cb?: ICallback<R, T>,
	options: ICloneRegexpOptionsCustom<T> = {},
)
{
	let ret: R[] = [];
	cb = cb ?? _cb as any

	for (const m of _each(input, re, options))
	{
		ret.push(cb(m.match, m.index, m.re, m))
	}

	return ret;
}

export default map
