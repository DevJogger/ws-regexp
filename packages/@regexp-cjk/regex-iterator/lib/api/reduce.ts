import { IChainArray } from '../types';
import { ICloneRegexpOptionsCustom, ICloneRegexpOptions } from '@regexp-cjk/clone-regexp/lib/types';
import { handleChainInput } from '../util/handleChainInput';
import { _each } from './_each';
import { handleChainRowResult } from '../util/handleChainRowResult';

export interface IOptionsReduceChain<T extends RegExp = RegExp> extends ICloneRegexpOptionsCustom<T>
{
	allowFallbackToSource?: boolean,
}

export function* _reduceCore(input: string, chain: IChainArray, options: IOptionsReduceChain = {})
{
	let { allowFallbackToSource } = options;

	chain = chain.slice();

	const { regexp: re, backref } = handleChainInput(chain.shift(), options);

	const options2: ICloneRegexpOptions = {
		...options,
		global: false,
	};

	for (const m of _each(input, re, options))
	{
		let str = handleChainRowResult({

			regexp: re,
			backref,

			match: m.match,

			allowFallbackToSource,
		}, m.match[0])

		if (chain.length > 0)
		{
			for (const row of chain)
			{
				let { regexp, backref } = handleChainInput(row, options2)

				const match = str.match(regexp);

				str = handleChainRowResult({
					regexp,
					backref,

					match,

					allowFallbackToSource,
				}, str)
			}
		}

		yield str
	}
}

export function reduce(input: string, chain: IChainArray, options: IOptionsReduceChain = {})
{
	return [..._reduceCore(input, chain, options)]
}

export default reduce
