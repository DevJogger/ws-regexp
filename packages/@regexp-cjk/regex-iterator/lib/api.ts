
import each from './api/each';
import map from './api/map';
import reduce from './api/reduce';
import replaceEach from './api/replaceEach';

export { each }
export { map }
export { reduce }
export { replaceEach }
