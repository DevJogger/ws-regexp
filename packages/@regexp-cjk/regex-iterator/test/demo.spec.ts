import map from '../lib/api/map';
import each from '../lib/api/each';
import reduce from '../lib/api/reduce';
import { replaceEach } from '../lib/api/replaceEach';

const re = /\S/g;
const input = 'ka kk'

test(`map`, () =>
{

	let actual = map(input, re);

	expect(actual).toHaveLength(4);
	expect(actual).toMatchSnapshot();

});

test(`map only once`, () =>
{

	let actual = map(input, /\S/);

	expect(actual).toHaveLength(1);
	expect(actual).toMatchSnapshot();

});

test(`each`, () =>
{
	expect.assertions(5)

	let actual = each(input, re, (m) =>
	{
		expect(m).toHaveLength(1);
	});

	expect(actual).toBeUndefined();

});

test(`each only once`, () =>
{

	expect.assertions(2)

	let actual = each(input, /\S/, (m) =>
	{
		expect(m).toHaveLength(1);
	});

	expect(actual).toBeUndefined();

});

test(`reduce`, () =>
{

	let html = '<a href="http://xregexp.com/api/">XRegExp</a>\
        <a href="http://www.google.com/">Google</a>';

	let actual = reduce(html, [
		{ regexp: /<a href="([^"]+)">/ig, backref: 1 },
		{ regexp: new RegExp('^https?://(?<domain>[^/?#]+)'), backref: 'domain' },
	])

	expect(actual).toStrictEqual(['xregexp.com', 'www.google.com']);

});

test(`reduce:throw`, () =>
{

	let html = '<a href="http://xregexp.com/api/">XRegExp</a>\
        <a href="http://www.google.com/">Google</a>';

	expect(() => reduce(html, [
		{ regexp: /<a href="([^"]+)">/ig, backref: 1 },
		{ regexp: new RegExp('^https?://(?<domain>[^/?#]+)'), backref: 'do' },
	])).toThrowErrorMatchingSnapshot()

});

test(`reduce:allowFallbackToSource`, () =>
{

	let html = '<a href="xregexp.com/api/">XRegExp</a>\
        <a href="http://www.google.com/">Google</a>';

	let actual = reduce(html, [
		{ regexp: /<a href="([^"]+)">/ig, backref: 1 },
		{ regexp: new RegExp('^https?://(?<domain>[^/?#]+)'), backref: 'domain' },
	], {
		allowFallbackToSource: true,
	})

	expect(actual).toStrictEqual(['xregexp.com/api/', 'www.google.com']);

	actual = reduce(html, [
		{ regexp: /<a href="([^"]+)">/ig, backref: 1 },
		{ regexp: new RegExp('^https?://(?<domain>[^/?#]+)'), backref: 'domain' },
	])

	expect(actual).toStrictEqual([void 0, 'www.google.com']);

});

test(`reduce only once`, () =>
{

	let html = '<a href="http://xregexp.com/api/">XRegExp</a>\
        <a href="http://www.google.com/">Google</a>';

	let actual = reduce(html, [
		{ regexp: /<a href="([^"]+)">/i, backref: 1 },
		{ regexp: new RegExp('^https?://(?<domain>[^/?#]+)'), backref: 'domain' },
	])

	expect(actual).toStrictEqual(['xregexp.com']);

});

test(`replaceEach`, () =>
{

	let actual = replaceEach('ka', [
		[/./g, 'u'],
		[/./, 'x'],
		{
			regexp: /u/,
			replaceValue: (s) => s + '777',
		}
	]);

	expect(actual).toStrictEqual('xu777');

});
