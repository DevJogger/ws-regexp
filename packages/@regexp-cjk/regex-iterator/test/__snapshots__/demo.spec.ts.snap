// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map 1`] = `
Array [
  Array [
    "k",
  ],
  Array [
    "a",
  ],
  Array [
    "k",
  ],
  Array [
    "k",
  ],
]
`;

exports[`map only once 1`] = `
Array [
  Array [
    "k",
  ],
]
`;

exports[`reduce:throw 1`] = `"Invalid regular expression: /^https?:\\\\/\\\\/(?<domain>[^/?#]+)/: Invalid named capture referenced 'do'"`;
